Follow up: deposit selection order for ReqSn

ffakenz opened the pull request:
<!-- Describe your change here -->

🔑 **Motivation** 
> Follow up on #2217

When multiple deposits become active at the same time, the node currently selects the next deposit for inclusion in `ReqSn` based on the `Ord` instance of `TxId`.  

This behavior is not intuitive from a user’s perspective and may lead to deposits being picked in an order unrelated to their actual arrival on-chain.  

To make the behavior predictable and fair, the selection should follow FIFO semantics, based on the `created` field of each deposit.

🔄 **Changes**

* Pending active deposits are now picked in arrival order (`created`) when building `ReqSn` in the `Open` state. 

📝 **Notes**

* This change ensures that users see deposits included in the order they were observed on-chain, improving transparency and reducing surprises.  

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [X] Documentation updated or not needed
* [X] Haddocks updated or not needed
* [X] No new TODOs introduced or explained herafter

ffakenz committed:
prove deposits are not picked in arrival order

ffakenz committed:
select next deposit for ReqSn using FIFO, based on deposit created

ffakenz committed:
update CHANGELOG

noonio commented on CHANGELOG.md:
```diff
@@ -42,6 +42,8 @@ changes.
 - **BREAKING** Enable handling client recover in all head states.
   - See [Issue #1812](https://github.com/cardano-scaling/hydra/issues/1812) and [PR #2217](https://github.com/cardano-scaling/hydra/pull/2217).
 
+- Select next deposit for ReqSn using FIFO, based on deposit created.
```
Please link to PR :)

vrom911 commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -979,14 +979,14 @@ onOpenChainTick env pendingDeposits st chainTime =
-    maybe noop (cont . fst) . find p $ Map.toList deposits
+    case filter p (Map.toList deposits) of
+      [] -> noop
+      xs -> cont (fst (minimumBy (\(_, Deposit{created = c1}) (_, Deposit{created = c2}) -> compare c1 c2) xs))
```
I think this could be simplified a bit visually:

```suggestion
      xs -> cont (fst (minimumBy (comparing (created . snd)) xs))
```
I think it still makes it nicely readable 🙏🏼

vrom911 replied:
Also, maybe adding a comment on the decision might help (either in the function doc or right here)

ffakenz replied:
Can't apply your suggestion directly due to: Ambiguous occurrence ‘created’.
> Please note this occurs even though the module already has `DuplicateRecordFields` enabled.

ffakenz committed:
comment and simplify withNextActive

noonio commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -979,14 +979,15 @@ onOpenChainTick env pendingDeposits st chainTime =
-    maybe noop (cont . fst) . find p $ Map.toList deposits
+    case filter p (Map.toList deposits) of
+      [] -> noop
+      xs -> cont (fst (minimumBy (comparing ((\Deposit{created} -> created) . snd)) xs))
```
Maybe: Could this map _already_ be sorted on insert? Then we just take the first one again?

vrom911 replied:
`Map.toList` doesn't guarantee that, I'm afraid. We could use `toDescList`, which probably would work. It is `O(n)` as well https://hackage-content.haskell.org/package/containers-0.8/docs/Data-Map-Internal.html#v:toDescList

noonio replied:
I suggested to @ffakenz to perhaps use [OMap instead](https://hackage.haskell.org/package/ordered-containers-0.2.4/docs/Data-Map-Ordered.html); not sure if the library is any good but might be promising!

ffakenz replied:
@noonio appreciate the suggestion!  

However, [OMap](https://hackage.haskell.org/package/ordered-containers-0.2.4/docs/Data-Map-Ordered.html) orders elements based on the `Ord` instance of the key, but in our case we need ordering by `created`.  Since `created` isn’t suitable as the key itself, OMap doesn’t quite fit here.  

@vrom911 same applies to your suggestion.

ffakenz replied:
For our scenario, because we want to **process items in the order of arrival**, the following data structures would be more suitable for strict FIFO behavior:

* [`Data.Sequence`](https://hackage-content.haskell.org/package/containers-0.8/docs/Data-Sequence.html)
* [`Data.Sequence.Internal.Sorting.Queue`](https://hackage-content.haskell.org/package/containers-0.8/docs/Data-Sequence-Internal-Sorting.html#t:Queue)

However, we also need to **look up deposits by key** in functions like `onOpenNetworkReqSn` and `onClientRecover`, and **delete pending deposits by key** during `aggregateNodeState`:

- [onOpenNetworkReqSn](https://github.com/cardano-scaling/hydra/blob/63046df0748787449c657c44461d01a16168285f/hydra-node/src/Hydra/HeadLogic.hs#L497): `Map.lookup depositTxId pendingDeposits`

- [onClientRecover](https://github.com/cardano-scaling/hydra/blob/63046df0748787449c657c44461d01a16168285f/hydra-node/src/Hydra/HeadLogic.hs#L759): `Map.lookup recoverTxId pendingDeposits`

- [aggregateNodeState](https://github.com/cardano-scaling/hydra/blob/63046df0748787449c657c44461d01a16168285f/hydra-node/src/Hydra/HeadLogic.hs#L1448): `Map.delete depositTxId pendingDeposits`

Given these requirements, **using a `Map` is simpler and more practical for now**, because it efficiently supports key-based lookup and deletion, even if it doesn’t strictly preserve FIFO order.

We could consider combining a queue (for strict FIFO) with a map later, but for the current logic, sticking with `Map` is the most straightforward approach.

noonio replied:
Fine

I suppose one could hack a little the Ord on the key by having a newtype and a specific instance; but I feel that's probably a bit too convoluted.

So let's leave it as-is.

Thanks for taking a look!

noonio commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -152,6 +153,39 @@ spec =
 
+      describe "Deposit" $ do
+        let plusTime = flip addUTCTime
+        it "on tick, picks the next active deposit in arrival when in Open state order for ReqSn" $ do
```
nice!

vrom911 reviewed (approved):
Lookin good 💯

ffakenz committed:
minor: fix exhaustive match in logic spec

ffakenz committed:
consolidate approach after rebase

we need to determine the next deposit status before we build ReqSn,
because events take effect after outcomes are aggregated.
